components/locid/src/parser/mod.rs

```diff
@@ -17,72 +17,91 @@ pub use locale::{
+    debug_assert!(idx < slice.len());
+
+    // This function is called only on the idx == 0 or on a separator.
+    let (start, mut end) = if is_separator(slice, idx) {
+        // If it's a separator, set the start to idx+1 and advance the idx to the next char.
+        (idx + 1, idx + 1)
+    } else {
+        // If it's idx=0, start is 0 and end is set to 1
+        debug_assert!(idx == 0);
+        (0, 1)
+    };
+
+    while end < slice.len() && !is_separator(slice, end) {
+        // Advance until we reach end of slice or a separator.
+        end += 1;
     }
+    // Notice: this slice may be empty (start == end) for cases like `"en-"` or `"en--US"`
+    (start, end)
 }
 
+// `SubtagIterator` is a helper iterator for [`LanguageIdentifier`] and [`Locale`] parsing.
+//
+// It is quite extraordinary due to focus on performance and Rust limitations for `const`
+// functions.
+//
+// The iterator is eager and fallible allowing it to reject invalid slices such as `"-"`, `"-en"`,
+// `"en-"` etc.
+//
+// The iterator provides methods available for static users - `next_manual` and `peek_manual`,
+// as well as typical `Peekable` iterator APIs - `next` and `peek`.
+//
+// All methods return an `Option` of a `Result`.
 #[derive(Copy, Clone, Debug)]
 pub struct SubtagIterator<'a> {
-    slice: &'a [u8],
-    current_start: usize,
-    current_end: usize,
+    pub slice: &'a [u8],
+    done: bool,
+    // done + subtag is faster than Option<(usize, usize)>
+    // at the time of writing.
+    subtag: (usize, usize),
 }
 
-pub type ManualSlice<'a> = (&'a [u8], usize, usize);
-
 impl<'a> SubtagIterator<'a> {
-    pub const fn next_manual(mut self) -> (Self, Option<ManualSlice<'a>>) {
-        if self.current_start == self.current_end {
-            (self, None)
-        } else {
-            let r = (self.slice, self.current_start, self.current_end);
-            self.current_start = self.current_end;
-            #[allow(clippy::indexing_slicing)]
-            while self.current_start < self.slice.len()
-                && (self.slice[self.current_start] == b'-'
-                    || self.slice[self.current_start] == b'_')
-            {
-                self.current_start += 1;
-            }
-            self.current_end = self.current_start;
-            #[allow(clippy::indexing_slicing)]
-            while self.current_end < self.slice.len()
-                && self.slice[self.current_end] != b'-'
-                && self.slice[self.current_end] != b'_'
-            {
-                self.current_end += 1;
-            }
-            (self, Some(r))
+    pub const fn try_new(slice: &'a [u8]) -> Result<Self, ParserError> {
+        // This returns early for slices like `"-en"` or `"-"`
+        if slice.is_empty() || is_separator(slice, 0) {
+            return Err(ParserError::InvalidLanguage);
         }
+        Ok(Self {
+            slice,
+            done: false,
+            subtag: get_current_subtag(slice, 0),
```

Comment:
Thought/Suggestion: What happens if you initialize the subtag to `(0, 0)` if the string starts with a separator; I think that should be sufficient for the rest of the code to correctly fail to match an empty subtag against the language? It might simplify your code since you don't need this function to return `Result`.

Reply:
It seems possible. Let me play with it and I'll file a separate PR.